### ISSUE TITLE
Super+. does not open emoji chooser in GTK 4 apps on Xorg

### ISSUE THREAD

**Issue by jtojnar** [ibus/ibus#2607]
### Distribution and version
NixOS unstable

### Desktop environment and version
GNOME 45

### Session type
Wayland (it works on X11)

### Which application and version?

gtk4-demo-application (GTK 4.12.4)

### IBus version

1.5.29

### Issue description

When I press <kbd>Super</kbd>-<kbd>.</kbd> in `gtk4-demo-application`, it inserts period instead of `e̲` placeholder. It works in `gtk3-demo-application`.

If I add another shortcut such as <kbd>Ctrl</kbd>-<kbd>Shift</kbd>-<kbd>e</kbd> in `ibus-setup`, that one will work correctly in GTK 4 apps too.

### Steps to reproduce

1. Run `gtk4-demo-application`
1. Press <kbd>Super</kbd>-<kbd>.</kbd>
1. `.` will be inserted, not emoji chooser placeholder

### Other details

GTK has built-in Emoji chooser bound to <kbd>Ctrl</kbd>-<kbd>.</kbd>. That one works in both GTK 3 and GTK 4 but I would prefer to use the IBus emoji chooser since that works in non-GTK apps too.

### Can you reproduce your problem when you restart ibus-daemon?

Yes

### Do you see any errors when you run ibus-daemon with the verbose option?

No

### Can you reproduce your problem with a new user account instead of the current your account?

Yes, used a fresh virtual machine instance.

**Comment by MURTOMAASORTAJA** [comment]
+1 from here for this.

I get the `e̲` placeholder even when the emoji thing is assigned to ctrl+period.

**Comment by fujiwarat** [comment]
> GNOME 45
> Wayland (it works on X11)

I think the result is reverse.
It works in GNOME Wayland with gnome-shell & mutter 45.4 but not in Xorg.

Can you attach the log of `dbus-monitor --address $(ibus address)` when you type Super-period?
I think gnome-shell & mutter sends the state 64 in Wayland but not Xorg:
```
method call time=1708005814.592375 sender=:1.3 -> destination=org.freedesktop.IBus serial=344 path=/org/freedesktop/IBus/InputContext_2; interface=org.freedesktop.IBus.InputContext; member=ProcessKeyEvent
   uint32 46
   uint32 52
   uint32 64

```

**Comment by jtojnar** [comment]
> I think the result is reverse.
> It works in GNOME Wayland with gnome-shell & mutter 45.4 but not in Xorg.

Not the case for me with 1.5.29. It works here in GTK 4 apps on Xorg but not on Wayland. But it works in GTK 3 apps on both.

|ibus 1.5.29| Wayland|Xorg|
|-|--|--|
|**GTK 3**|✅|✅|
|**GTK 4**|❌|✅|

- [ibus-1.5.29-gtk3-wayland.txt](https://github.com/ibus/ibus/files/14324665/ibus-1.5.29-gtk3-wayland.txt)
- [ibus-1.5.29-gtk3-x11.txt](https://github.com/ibus/ibus/files/14324666/ibus-1.5.29-gtk3-x11.txt)
- [ibus-1.5.29-gtk4-wayland.txt](https://github.com/ibus/ibus/files/14324667/ibus-1.5.29-gtk4-wayland.txt)
- [ibus-1.5.29-gtk4-x11.txt](https://github.com/ibus/ibus/files/14324668/ibus-1.5.29-gtk4-x11.txt)

The behaviour with 1.5.28 is closer to what you describe:

|ibus 1.5.28| Wayland|Xorg|
|-|--|--|
|**GTK 3**|✅|❌|
|**GTK 4**|❌|❌|

- [ibus-1.5.28-gtk3-x11.txt](https://github.com/ibus/ibus/files/14324672/ibus-1.5.28-gtk3-x11.txt)
- [ibus-1.5.28-gtk4-wayland.txt](https://github.com/ibus/ibus/files/14324673/ibus-1.5.28-gtk4-wayland.txt)
- [ibus-1.5.28-gtk4-x11.txt](https://github.com/ibus/ibus/files/14324674/ibus-1.5.28-gtk4-x11.txt)
- [ibus-1.5.28-gtk3-wayland.txt](https://github.com/ibus/ibus/files/14324671/ibus-1.5.28-gtk3-wayland.txt)

**Comment by jtojnar** [comment]
Trying 1.5.30-beta1, it behaves the same as 1.5.29:

|ibus 1.5.30-beta1| Wayland|Xorg|
|-|--|--|
|**GTK 3**|✅|✅|
|**GTK 4**|❌|✅|

- [ibus-1.5.30-beta1-gtk4-wayland.txt](https://github.com/ibus/ibus/files/14324813/ibus-1.5.30-beta1-gtk4-wayland.txt)
- [ibus-1.5.30-beta1-gtk3-x11.txt](https://github.com/ibus/ibus/files/14324812/ibus-1.5.30-beta1-gtk3-x11.txt)
- [ibus-1.5.30-beta1-gtk3-wayland.txt](https://github.com/ibus/ibus/files/14324811/ibus-1.5.30-beta1-gtk3-wayland.txt)
- [ibus-1.5.30-beta1-gtk4-x11.txt](https://github.com/ibus/ibus/files/14324814/ibus-1.5.30-beta1-gtk4-x11.txt)

Still do not see `uint32 64` anywhere.

**Comment by fujiwarat** [comment]
> * ibus-1.5.30-beta1-gtk4-wayland.txt
```
method call time=1708297864.799991 sender=:1.9 -> destination=org.freedesktop.IBus serial=24 path=/org/freedesktop/IBus/InputContext_6; interface=org.freedesktop.IBus.InputContext; member=ProcessKeyEvent
   uint32 46
   uint32 52
   uint32 67108864
```

Your gnome-shell/mutter sends the different values to ibus.

**Comment by fujiwarat** [comment]
I guess I and you have the different build configurations or versions as I noted the versions above.
Since our results are reverse, I cannot say why your problem happens.
I will check GNOME 46 in Fedora 40 later.

**Comment by fujiwarat** [comment]
In my tests, I observe `GDK_SUPER_MASK` from /usr/include/gtk-4.0/gdk/gdkenums.h with GTK4 applications in Xorg and `GDK_MOD4_MASK` from /usr/include/gtk-3.0/gdk/gdktypes.h with GTK4 applications in Wayland.
I guess gnome-shell/mutter still refers the GTK3 values in Wayland and GTK4 refers GTK4 values in Xorg but your results are reverse.

**Comment by fujiwarat** [comment]
I tried this issue with gnome-shell-46~rc, mutter-46~rc, gtk4-4.13.8 but the result is still reverse.

I guess your environment has a wrong value in GSettings `gtk-im-module` in the GNOME Wayland session.
```
% settings get org.gnome.desktop.interface gtk-im-module
''
```

**Comment by fujiwarat** [comment]
I tried to fix the issue in the client side so I needed to know which client GTK4 or Mutter effect this issue.
However now I decided to fix the issue in the IBus engine side and treat both `GDK_MOD4_MASK` and `GDK_SUPER_MASK` at the moment and this way can cover even if the test result is reverse.

**Comment by fujiwarat** [comment]
Seems the previous fix causes a regression in 1.5.30-rc2.
